File: app/src/lib/apa.ts
/**
 * Render the engine's structured APA output (RichText runs, ApaTable cells) to plain text and to
 * self-contained HTML for the clipboard (SPEC §10.1, §10.3): italic symbols, no vertical rules,
 * horizontal rules above/below the header and at the bottom, italic title under a bold
 * "Table N" line, and "Note." notes. The engine already formatted every number.
 */
import type { ApaColumn, ApaTable, RichText, TableCell, TextRun } from "@/contracts";

export function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}

export function richToPlain(runs: RichText | null | undefined): string {
  return (runs ?? []).map((r) => r.text).join("");
}

function runToHtml(r: TextRun): string {
  let h = escapeHtml(r.text);
  if (r.subscript) h = `<sub>${h}</sub>`;
  if (r.superscript) h = `<sup>${h}</sup>`;
  if (r.italic) h = `<i>${h}</i>`;
  return h;
}

export function richToHtml(runs: RichText | null | undefined): string {
  return (runs ?? []).map(runToHtml).join("");
}

export function cellToPlain(c: TableCell): string {
  switch (c.type) {
    case "number":
    case "p_value":
    case "interval":
      return c.display;
    case "text":
      return richToPlain(c.text);
    default:
      return "";
  }
}

export function cellToHtml(c: TableCell): string {
  return c.type === "text" ? richToHtml(c.text) : escapeHtml(cellToPlain(c));
}

/** CSS text-align for a column; "decimal" falls back to right alignment in HTML/Word. */
export function alignOf(col: ApaColumn): "left" | "center" | "right" {
  return col.align === "decimal" ? "right" : col.align;
}

export function tableNumberLabel(t: ApaTable, fallback = 1): string {
  return `Table ${t.number ?? fallback}`;
}

/** Plain-text rendering (tab-separated) for the clipboard's text/plain flavour. */
export function tableToPlain(t: ApaTable, fallbackNumber = 1): string {
  const lines = [tableNumberLabel(t, fallbackNumber), t.title];
  lines.push(t.columns.map((c) => richToPlain(c.header)).join("\t"));
  for (const row of t.rows) {
    const cells = row.cells.map(cellToPlain);
    if (row.indent) cells[0] = "  ".repeat(row.indent) + cells[0];
    lines.push(cells.join("\t"));
  }
  const notes = noteParts(t);
  if (notes.length) lines.push(notes.map(richToPlain).join(" "));
  return lines.join("\n");
}

function noteParts(t: ApaTable): RichText[] {
  const parts: RichText[] = [];
  if (t.notes.general) parts.push([{ text: "Note.", italic: true }, { text: " " }, ...t.notes.general]);
  parts.push(...t.notes.specific, ...t.notes.probability);
  return parts;
}

const RULE = "1px solid #000";

/** Self-contained APA 7 table HTML with inline styles (pastes into Word with formatting). */
export function tableToHtml(t: ApaTable, fallbackNumber = 1): string {
  const font = "font-family:'Times New Roman',Times,serif;font-size:12pt;";
  const pad = "padding:2pt 8pt;";
  const groupRow = t.column_groups.length
    ? (() => {
        const cells: string[] = [];
        let col = 0;
        for (const g of [...t.column_groups].sort((a, b) => a.first_column - b.first_column)) {
          if (g.first_column > col) cells.push(`<th colspan="${g.first_column - col}" style="${pad}"></th>`);
          cells.push(
            `<th colspan="${g.span}" style="${pad}text-align:center;font-weight:normal;border-bottom:${RULE};">${richToHtml(g.label)}</th>`,
          );
          col = g.first_column + g.span;
        }
        if (col < t.columns.length) cells.push(`<th colspan="${t.columns.length - col}" style="${pad}"></th>`);
        return `<tr>${cells.join("")}</tr>`;
      })()
    : "";
  const header = t.columns
    .map((c) => `<th style="${pad}text-align:${alignOf(c)};font-weight:normal;border-bottom:${RULE};">${richToHtml(c.header)}</th>`)
    .join("");
  const body = t.rows
    .map((row, ri) => {
      const last = ri === t.rows.length - 1;
      const cells = row.cells
        .map((c, ci) => {
          const col = t.columns[ci];
          const indent = ci === 0 && row.indent ? `padding-left:${8 + row.indent * 12}pt;` : "";
          const weight = row.kind === "section_header" ? "font-style:italic;" : "";
          const bottom = last ? `border-bottom:${RULE};` : "";
          return `<td style="${pad}${indent}${weight}${bottom}text-align:${col ? alignOf(col) : "left"};">${cellToHtml(c)}</td>`;
        })
        .join("");
      return `<tr>${cells}</tr>`;
    })
    .join("");
  const notes = noteParts(t)
    .map((n) => `<p style="${font}margin:4pt 0 0 0;">${richToHtml(n)}</p>`)
    .join("");
  return (
    `<div style="${font}">` +
    `<p style="${font}margin:0;font-weight:bold;">${escapeHtml(tableNumberLabel(t, fallbackNumber))}</p>` +
    `<p style="${font}margin:0 0 6pt 0;font-style:italic;">${escapeHtml(t.title)}</p>` +
    `<table style="${font}border-collapse:collapse;border-top:${RULE};border-bottom:${RULE};">` +
    `<thead>${groupRow}<tr>${header}</tr></thead><tbody>${body}</tbody></table>${notes}</div>`
  );
}

export interface CopyPayload {
  html: string;
  text: string;
}

/** Rich + plain clipboard payload for an APA sentence. */
export function sentencePayload(runs: RichText): CopyPayload {
  return {
    html: `<p style="font-family:'Times New Roman',Times,serif;font-size:12pt;">${richToHtml(runs)}</p>`,
    text: richToPlain(runs),
  };
}

export function tablePayload(t: ApaTable, fallbackNumber = 1): CopyPayload {
  return { html: tableToHtml(t, fallbackNumber), text: tableToPlain(t, fallbackNumber) };
}

// --- display formatting for values the engine did not pre-format (SPEC §10.1) --------------

/** Effect sizes / statistics that cannot exceed 1 in magnitude: no leading zero in APA. */
const BOUNDED = /^(r|rho|tau.*|r_pb|rank_biserial|.*eta.*|omega.*|epsilon.*|kendall_w|cramers_v|phi|r_sq|alpha|omega|.*_r)$/;

export function isBounded(key: string): boolean {
  return BOUNDED.test(key);
}

export function fmtNum(x: number | null | undefined, digits = 2, bounded = false): string {
  if (x === null || x === undefined || !Number.isFinite(x)) return "—";
  const s = x.toFixed(digits);
  return bounded ? s.replace(/^(-?)0\./, "$1.") : s;
}

/** APA p: three decimals, no leading zero, "< .001" floor. Returns the text after "p". */
export function fmtPExpr(p: number | null | undefined): string {
  if (p === null || p === undefined || !Number.isFinite(p)) return "= —";
  if (p < 0.001) return "< .001";
  return `= ${p.toFixed(3).replace(/^0\./, ".")}`;
}

export function fmtDf(df: number[]): string {
  return df.map((d) => (Number.isInteger(d) ? String(d) : d.toFixed(2))).join(", ");
}

File: engine/statly_engine/stats/apa.py
"""APA 7 number formatting and styled text runs (SPEC §10.1).

Rules: italic statistical symbols; values that cannot exceed 1 in magnitude (p, r, alpha,
correlations, proportions of variance) have no leading zero; p to three decimals with a
"p < .001" floor; other statistics to two decimals. The engine is the single APA formatter:
tables carry these display strings, the frontend never re-formats numbers.
"""

from __future__ import annotations

import math

MINUS = "-"  # hyphen-minus: pastes cleanly into Word, Excel and plain text
EM_DASH = "—"  # APA placeholder for an empty / undefined cell


def finite(x) -> bool:
    return x is not None and isinstance(x, (int, float)) and math.isfinite(x)


def num(x, decimals: int = 2) -> str:
    """Fixed decimals with a leading zero (M, SD, t, F, d...). Non-finite -> em dash."""
    if not finite(x):
        return EM_DASH
    s = f"{x:.{decimals}f}"
    if s.startswith("-") and float(s) == 0:
        s = s[1:]  # no "-0.00"
    return s.replace("-", MINUS)


def no_zero(x, decimals: int = 2) -> str:
    """No leading zero, for values bounded by 1 (r, alpha, eta squared): .34, -.05, 1.00."""
    s = num(x, decimals)
    if s.startswith("0."):
        return s[1:]
    if s.startswith(MINUS + "0."):
        return MINUS + s[2:]
    return s


def integer(x) -> str:
    return str(int(x)) if finite(x) else EM_DASH


def df_text(df) -> str:
    """Integer df print without decimals; fractional (Welch, GG) df to two decimals."""
    if not finite(df):
        return EM_DASH
    return str(int(round(df))) if abs(df - round(df)) < 1e-9 else num(df, 2)


def p_value(p) -> str:
    """Table display: '.034', '< .001', '> .999'."""
    if not finite(p):
        return EM_DASH
    if p < 0.001:
        return "< .001"
    if p > 0.999:
        return "> .999"
    return no_zero(p, 3)


def p_relation(p) -> str:
    """In-text form after the italic p: ' = .034' or ' < .001'."""
    if not finite(p):
        return " = " + EM_DASH
    s = p_value(p)
    return " " + s if s[0] in "<>" else " = " + s


def ci_text(lower, upper, decimals: int = 2, bounded: bool = False) -> str:
    """'[0.12, 0.88]'. Open (one-sided) bounds print as -∞ / ∞."""
    fmt = no_zero if bounded else num
    lo = fmt(lower, decimals) if finite(lower) else MINUS + "∞"
    hi = fmt(upper, decimals) if finite(upper) else "∞"
    return f"[{lo}, {hi}]"


# ---------------------------------------------------------------------------
# Rich text (contracts AnalysisResult#/$defs/RichText)
# ---------------------------------------------------------------------------
class Rich:
    """Builder for a RichText run list. Adjacent runs with identical styling are merged.

    >>> Rich().i("t").t("(28) = 2.10, ").i("p").t(" = .045").runs
    """

    def __init__(self):
        self.runs: list[dict] = []

    def _add(self, text: str, italic=False, subscript=False, superscript=False) -> "Rich":
        if not text:
            return self
        run = {"text": text, "italic": italic, "subscript": subscript, "superscript": superscript}
        last = self.runs[-1] if self.runs else None
        if last and all(last[k] == run[k] for k in ("italic", "subscript", "superscript")):
            last["text"] += text
        else:
            self.runs.append(run)
        return self

    def t(self, text: str) -> "Rich":
        return self._add(text)

    def i(self, text: str) -> "Rich":
        return self._add(text, italic=True)

    def sub(self, text: str, italic: bool = False) -> "Rich":
        return self._add(text, italic=italic, subscript=True)

    def sup(self, text: str) -> "Rich":
        return self._add(text, superscript=True)

    def extend(self, other: "Rich | list[dict]") -> "Rich":
        for r in (other.runs if isinstance(other, Rich) else other):
            self._add(r["text"], r.get("italic", False), r.get("subscript", False), r.get("superscript", False))
        return self

    def symbol(self, sym: "str | Rich") -> "Rich":
        """Italic symbol; a Rich passes through (e.g. d with subscript 'av')."""
        return self.extend(sym) if isinstance(sym, Rich) else self.i(sym)

    def stat(self, sym, df: list | tuple, value, decimals: int = 2) -> "Rich":
        """'t(28.51) = 2.10' with italic symbol."""
        self.symbol(sym)
        if df:
            self.t("(" + ", ".join(df_text(d) for d in df) + ")")
        return self.t(" = " + num(value, decimals))

    def p(self, p) -> "Rich":
        return self.i("p").t(p_relation(p))

    def es(self, sym, value, lower, upper, level: float, bounded: bool = False) -> "Rich":
        """'d = 0.52, 95% CI [0.10, 0.94]'."""
        fmt = no_zero if bounded else num
        self.symbol(sym).t(" = " + fmt(value, 2))
        if finite(lower) or finite(upper):
            self.t(f", {level_text(level)} CI {ci_text(lower, upper, 2, bounded)}")
        return self

    def plain(self) -> str:
        return "".join(r["text"] for r in self.runs)


def level_text(level: float) -> str:
    pct = level * 100
    return f"{pct:.0f}%" if abs(pct - round(pct)) < 1e-9 else f"{pct:g}%"


def text(s: str) -> list[dict]:
    return Rich().t(s).runs


def italic(s: str) -> list[dict]:
    return Rich().i(s).runs


# ---------------------------------------------------------------------------
# Table cells (contracts AnalysisResult#/$defs/TableCell)
# ---------------------------------------------------------------------------
def cell_text(s: "str | Rich | list[dict]") -> dict:
    runs = s.runs if isinstance(s, Rich) else (text(s) if isinstance(s, str) else s)
    return {"type": "text", "text": runs}


def cell_num(x, decimals: int = 2, bounded: bool = False) -> dict:
    v = float(x) if finite(x) else None
    return {"type": "number", "value": v, "display": (no_zero if bounded else num)(x, decimals)}


def cell_int(x) -> dict:
    return {"type": "number", "value": float(x) if finite(x) else None, "display": integer(x)}


def cell_df(x) -> dict:
    return {"type": "number", "value": float(x) if finite(x) else None, "display": df_text(x)}


def cell_p(p) -> dict:
    return {"type": "p_value", "value": float(p) if finite(p) else None, "display": p_value(p)}


def cell_ci(lower, upper, decimals: int = 2, bounded: bool = False) -> dict:
    return {"type": "interval", "lower": float(lower) if finite(lower) else None,
            "upper": float(upper) if finite(upper) else None, "display": ci_text(lower, upper, decimals, bounded)}


def cell_empty() -> dict:
    return {"type": "empty"}


def column(key: str, header: "str | Rich", align: str = "decimal") -> dict:
    runs = header.runs if isinstance(header, Rich) else text(header)
    return {"key": key, "header": runs, "align": align}


def row(cells: list[dict], indent: int = 0, kind: str = "data") -> dict:
    return {"cells": cells, "indent": indent, "kind": kind}


def table(title: str, columns: list[dict], rows: list[dict], *, number: int | None = 1,
          column_groups: list[dict] | None = None, general_note: "Rich | str | None" = None,
          specific_notes: list | None = None, probability_notes: list | None = None) -> dict:
    """APA table: numbered, italic title (renderer italicizes), horizontal rules only, notes."""
    def rt(x):
        return x.runs if isinstance(x, Rich) else text(x)

    return {
        "number": number, "title": title, "columns": columns, "column_groups": column_groups or [],
        "rows": rows,
        "notes": {"general": rt(general_note) if general_note is not None else None,
                  "specific": [rt(n) for n in specific_notes or []],
                  "probability": [rt(n) for n in probability_notes or []]},
    }


def column_group(label: "str | Rich", first_column: int, span: int) -> dict:
    return {"label": label.runs if isinstance(label, Rich) else text(label), "first_column": first_column,
            "span": span}
